bayes_kit/drghmc.py

```diff
@@ -0,0 +1,437 @@
+from typing import Iterator, Optional, Union, Callable
+
+import numpy as np
+
+from .typing import DrawAndLogP, GradModel, Seed, VectorType
+
+
+class DrGhmcDiag:
+    """Generalized HMC sampler with (Probabilistic) Delayed Rejection diagonal metric.
+
+    To sample from a target distribution, DrGhmcDiag proposes a new sample from the
+    current sample using the leapfrog integrator. The probability of accepting this
+    transition is recursively computed. If accepted, DrGhmcDiag returns the proposed
+    sample and its log density. If rejected, DrGhmcDiag proposes a new sample generated
+    with new stepsize and number of steps for the leapfrog integrator. This process is
+    repeated until a maximum number of proposals is reached or probabilistic delayed
+    rejection terminates early.
+
+    Intuitively, DrGhmcDiag reduces the stepsize in subsequent proposals if the
+    previous proposal was rejected for better sampling resolution of high-curvature
+    state space regions.
+
+    Implementation based on Modi, Chirag, Alex Barnett, and Bob Carpenter. "Delayed
+    rejection Hamiltonian Monte Carlo for sampling multiscale distributions." Bayesian
+    Analysis 1.1 (2023): 1-28. While the paper represents position and momentum with
+    (p, q), this implementation uses (theta, rho) for consistency with bayes-kit. All
+    computation is done on the log scale unless explicitly stated otherwise.
+    """
+
+    def __init__(
+        self,
+        model: GradModel,
+        stepsize: Union[int | float, list[int | float], Callable[[int], int | float]],
```

Comment:
Using `|` as a synonym for `Union` isn't available in Python 3.9.

I'm not sure if this is related to the failures mypy is raising or not

Reply:
I will change `|` to `Union`.

I originally made this change because of a recommendation-in-passing I saw in Google's style doc here: https://google.github.io/styleguide/pyguide.html#3195-nonetype.